pkg/kube/kclient/client_test.go

```diff
@@ -350,6 +350,45 @@ func TestClient(t *testing.T) {
 	assert.Equal(t, tester.Get(obj3.Name, obj3.Namespace), nil)
 }
 
+func TestShutdown(t *testing.T) {
+	tracker := assert.NewTracker[string](t)
+	removeTracker := assert.NewTracker[string](t)
+	c := kube.NewFakeClient()
+	deployments := kclient.NewFiltered[*appsv1.Deployment](c, kclient.Filter{})
+	deployments.AddEventHandler(clienttest.TrackerHandler(tracker))
+	removeReg := deployments.AddEventHandler(clienttest.TrackerHandler(removeTracker))
+	tester := clienttest.Wrap(t, deployments)
+
+	c.RunAndWait(test.NewStop(t))
+	obj1 := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{Name: "1", Namespace: "default"},
+		Spec:       appsv1.DeploymentSpec{MinReadySeconds: 1},
+	}
+
+	// Create object, make sure we can see it
+	tester.Create(obj1)
+	// Client is cached, so its only eventually consistent
+	tracker.WaitUnordered("add/1")
+	removeTracker.WaitUnordered("add/1")
+
+	// Shutdown one
+	deployments.ShutdownHandler(removeReg)
+
+	// Update object, should see the update...
+	obj1.Spec.MinReadySeconds = 2
+	tester.Update(obj1)
+	tracker.WaitOrdered("update/1")
+	removeTracker.Empty()
+
+	deployments.ShutdownHandler(removeReg)
+
+	// Update object, should see the update...
+	obj1.Spec.MinReadySeconds = 2
+	tester.Update(obj1)
+	tracker.Empty()
+	removeTracker.Empty()
```

Comment:
What is the second ShutdownHandler and then doing a noop update testing?

Reply:
A badly written test. I fixed it and improved the comments too. Good catch